### ISSUE TITLE
Layer name propagation to backend

### ISSUE THREAD

**Issue by mirosval** [fchollet/keras#3547]
I want to use Keras as a high-level front for Tensorflow. I export the TF graph, reduce it only to the inference part and then deploy it.

The problem is that I want the names of nodes in the TF graph to be deterministic. i.e. if I create a keras layer `Activation('softmax', name="output")` I want there to be a node named `output` in the TF graph. (or at least something like `softmax_output_1`)

This does not happen, because Keras backend `softmax(x)` does not take a name parameter. I presume that the reason behind this is that while TF does allow you to name your nodes, theano does not (or at least it seemed so when I checked theano softmax). 

I understand that this might not be something you want, i.e. being able to name nodes only in one backend if the other does not support it, however it seems very useful and should not actually harm theano users. 

I would like to contribute this functionality if it is not something you totally disagree with.

**Comment by fchollet** [comment]
Keras is compatible with TF name scopes. Maybe this can help you achieve
what you want?

**Comment by mirosval** [comment]
Generally that's a very good idea, but isn't that limited to the model definition run? i.e. if I serialise and unserialise the model later doesn't it get lost?

I'd like to have a pipeline like this:

1. Create/Poke around model with Keras in Jupyter
2. Serialise model & weights
2. (Optional) Train model automatically in production & update the weights
3. Convert it into a minimal TF graph by removing all the training nodes
3. Serialise using TF protobuf
4. Load & Serve from Scala

**Comment by ahundt** [comment]
Would the direct use of tf scoping necessarily make a given model incompatible with the theano backend?

#3357 looks relevant to this

**Comment by fchollet** [comment]
We should create a `K.name_scope` operation which should also work for
Theano variable names. Then use this op in layers (e.g. in __call__) to
make sure all ops have appropriate names.

Feel free to open a PR.

**Comment by madhavajay** [comment]
This is still a problem, does anyone have a solution to this? I cannot figure out what my output node name is in the pbtxt file. I am currently compiling summarize_graph as per the TensorFlow documentation, but I would love to be able to set this in Keras.
